modules/xcm-bridge-hub-router/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright 2019-2021 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Pallet that may be used instead of `SovereignPaidRemoteExporter` in the XCM router
+//! configuration. The main thing that the pallet offers is the dynamic message fee,
+//! that is computed based on the bridge queues state. It starts exponentially increasing
+//! if the queue between this chain and the sibling/child bridge hub is overloaded.
+//!
+//! All other bridge hub queues offer some backpressure mechanisms. So if at least one
+//! of all queues is overloaded, it will eventually lead to the growth of the queue at
+//! this chain.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use bp_xcm_bridge_hub_router::LocalXcmQueue;
+use codec::Encode;
+use frame_support::traits::Get;
+use sp_runtime::{traits::One, FixedPointNumber, FixedU128, Saturating};
+use xcm::prelude::*;
+use xcm_builder::{ExporterFor, SovereignPaidRemoteExporter};
+
+pub use pallet::*;
+
+/// The factor that is used to increase current message fee factor when bridge experiencing
+/// some lags.
+const EXPONENTIAL_FEE_BASE: FixedU128 = FixedU128::from_rational(105, 100); // 1.05
+/// The factor that is used to increase current message fee factor for every sent kilobyte.
+const MESSAGE_SIZE_FEE_BASE: FixedU128 = FixedU128::from_rational(1, 1000); // 0.001
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config<I: 'static = ()>: frame_system::Config {
+		/// Universal location of this runtime.
+		type UniversalLocation: Get<InteriorMultiLocation>;
+		/// Relative location of the sibling bridge hub.
+		type SiblingBridgeHubLocation: Get<MultiLocation>;
+		/// The bridged network that this config is for.
+		type BridgedNetworkId: Get<NetworkId>;
+
+		/// Actual message sender (XCMP/DMP) to the sibling bridge hub location.
+		type ToBridgeHubSender: SendXcm + LocalXcmQueue;
```

Comment:
Need to split it. Something like that:
```rust
type ToBridgeHubSender: SendXcm;
type LocalXcmQueue: IsChannelActive;
```

And also use similar `IsChannelActive` in the message dispatch implementation. Probably we need to add 2 methods to cumulus XCMP pallet that will return `true` if inbound/outbound channel is active and provide some easy adapter for implementing `IsChannelActive` using functions that return `bool`.

Reply:
First part done